src/number_theory/sum_four_squares.lean

```diff
@@ -0,0 +1,215 @@
+/-
+Copyright (c) 2019 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+
+The main result in this file is `sum_four_squares`,
+  a proof that every natural number is the sum of four square numbers.
+-/
+import data.zmod.basic field_theory.finite group_theory.perm.sign
+import data.int.parity
+
+open finset polynomial finite_field equiv
+
+namespace int
+
+lemma sum_two_squares_of_two_mul_sum_two_squares {m x y : ℤ} (h : 2 * m =  x^2 + y^2) :
+  m = ((x - y) / 2) ^ 2 + ((x + y) / 2) ^ 2 :=
+have (x^2 + y^2).even, by simp [h.symm, even_mul],
+have hxaddy : (x + y).even, by simpa [pow_two] with parity_simps,
+have hxsuby : (x - y).even, by simpa [pow_two] with parity_simps,
+have (x^2 + y^2) % 2 = 0, by simp [h.symm],
+(domain.mul_left_inj (show (2*2 : ℤ) ≠ 0, from dec_trivial)).1 $
+calc 2 * 2 * m = (x - y)^2 + (x + y)^2 : by rw [mul_assoc, h]; ring
+... = (2 * ((x - y) / 2))^2 + (2 * ((x + y) / 2))^2 :
+  by rw [int.mul_div_cancel' hxsuby, int.mul_div_cancel' hxaddy]
+... = 2 * 2 * (((x - y) / 2) ^ 2 + ((x + y) / 2) ^ 2) :
+  by simp [mul_add, _root_.pow_succ, mul_comm, mul_assoc, mul_left_comm]
+
+lemma exists_sum_two_squares_add_one_eq_k {p : ℕ} (hp : p.prime) (hp1 : p % 2 = 1) :
+  ∃ (a b : ℤ) (k : ℕ), a^2 + b^2 + 1 = k * p ∧ k < p :=
```

Comment:
If `p % 2 = 0` I can take `a = 0`, `b = 1` and `k = 1`. So you don't need `hp1`, right?

Reply:
Yes. This actually makes the proof a little nicer, and the long inductive proof works for all primes, not just odd primes, so I don't even need to explicitly prove that two is the sum of four squares, although I basically did it here.